spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java

```diff
@@ -477,11 +486,12 @@ private void removeEmptyGroupAfterTimeout(MessageGroup messageGroup, long timeou
 							boolean removeGroup = groupNow.size() == 0 &&
 									groupNow.getLastModified()
 											<= (System.currentTimeMillis() - this.minimumTimeoutForEmptyGroups);
-							if (removeGroup) {
+							if (removeGroup && this.groupIds.contains(groupId)) {
```

Comment:
See my comment for the `remove()`  method changes. 
This should not be changed at all.
Although I would see `remove(groupNow)` call instead of direct `removeMessageGroup()`